### DISCUSSION TITLE
Why is there no normal categorical cross-entropy loss?

### DISCUSSION THREAD

**Discussion by yasharhon** [google/flax#4668] [Q&A] [answered]
This question concerns Optax just as much as Flax, but as the issue seems to stem from expections from Flax models, I am asking it here.

I was following [this example](https://docs.jaxstack.ai/en/latest/getting_started_with_jax_for_AI.html) in the JAX AI stack documentation, and noticed something odd. When implementing a classifier for images of digits, the model does not by default output softmaxxed logits. Instead, softmaxxing is invoked when calling `optax.softmax_cross_entropy_with_integer_labels`.

Given that `optax` does not even include a normal categorical cross entropy (between predicted and true labels), I assume it is expected that a JAX model for a classifier always outputs logits. But softmaxxing ensures that my model output conforms to the properties I would expect of a probabilistic classifier, and the model is trained as if the output is softmaxxed through the loss function. So is there a reason why such a model should be expected to output logits?

**Comment by carlosgmartin** [comment, marked as answer]
Computing cross-entropy directly from logits rather than probabilities avoids some redundant computation and/or improves numerical accuracy.

PyTorch takes a similar approach (see [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) and [NLLLoss](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html)), so this isn't unique to optax or flax.
